### ISSUE TITLE
Scripting

### ISSUE THREAD

**Issue by thekid** [xp-framework/rfc#332]
Scope of Change
==
This RFC groups multiple enhancements to the scripting facilities in the XP Framework

Rationale
==

Functionality
==
Scripts are declared in files with the `.script.php` file extension. As a short alternative for a class with a static *main()* method, they contain code directly at top-level.

Example
---------
```php
<?php namespace test;  // In a file test.script.php

use util\cmd\Console;

Console::writeLine('Hello ', $argv[1] ?? getenv('USER'));
```

Scripts are executed by the XP runner by `xp test.script.php`. Command line arguments are accessible via `$argv`.

Dependencies
---------------
The `Console` class in the above example comes from *xp-framework/core*, which is always available. In order to load classes from other libraries, scripts can reference them with the `from` extension to the `use` statement:

```php
<?php namespace test;

use net\daringfireball\markdown\Markdown from 'xp-forge/markdown';
```

This was introduced in xp-framework/core#161 and loads the library from Composer's global requirements.

🆕  User modules
---------------------
The first extension to the dependency loading mechanism is to include support for a user- and namespace specific directory instead of just Composer's global dependencies. It would search libraries in *$APPDATA/Xp/{namespace}* (Windows), *$XDG_CONFIG_HOME/xp/{namespace}* (XDG) and *$HOME/.xp/{namespace}* (other).

In the above example, the *xp-forge/markdown* library would be searched for in `~/.xp/test/vendor/xp-forge/markdown` in addition to the global Composer repository.

See xp-framework/core#207

🆕 User module versioning
------------------------------
The second extension would take module loading a step further and include version numbers and/or constraints, much like inside a *composer.json* require-statement.

```php
<?php namespace test;

use net\daringfireball\markdown\Markdown from 'xp-forge/markdown@^5.0';
```

Because module versions can affect the XP Framework version, the script's class path would need to determined by the XP Runners. Given the aforementioned convention, they could simply set it to `~/.xp/test/vendor/autoload.php`...

See xp-runners/reference#76 

🆕  Filenames
----------------
Currently error messages and stack traces do not include the line number of the file, but rather a reference to the executor, the *Code* class, e.g.:

```
at <native>::ParseError(0, (0x2a)'syntax error, unexpected '??' (T_COALESCE)')
   [line 1 of Code.class.php(127) : eval()'d code]
```

This will be addressed by using stream wrappers to load the files, like in xp-framework/compiler#54.

See xp-framework/core#208

Security considerations
==
n/a

Speed impact
==
Stat call for another directory, not noticeable

Dependencies
==
Change to XP Runners

Related documents
==
https://github.com/Kotlin/KEEP/blob/master/proposals/scripting-support.md - Kotlin Scripting
https://github.com/holgerbrandl/kscript - kscript - Having fun with Kotlin scripting
https://github.com/oleg-shilo/cs-script - CS-Script: C# scripting platform
https://github.com/dotnet/roslyn/wiki/Scripting-API-Samples - Roslyn Scripting API Samples
https://github.com/erning/gorun - gorun is a tool enabling one to put a "bang line" in the source code of a Go program to run it,

**Comment by thekid** [comment]
Before:

 ```bash
$ xp test.script.php
# Use path: first-to-exist[
#  $USE_XP
#  xp.ini:use_xp
#  ./vendor/xp-framework/core
#  $APPDATA/Composer/vendor/xp-framework/core
#]
```

After:

 ```bash
$ xp test.script.php
# Use path: first-to-exist[
#  $USER/test/vendor/xp-framework/core
#  $USE_XP
#  xp.ini:use_xp
#  ./vendor/xp-framework/core
#  $APPDATA/Composer/vendor/xp-framework/core
#]
``````

**Comment by thekid** [comment]
The *user module versioning* part should be implemented ontop of the changes from xp-framework/core#207 - this is why there is currently no pull request for this yet.

**Comment by thekid** [comment]
PR xp-framework/core#208 has been merged into a bugfix release, 9.8.2

**Comment by thekid** [comment]
PR xp-framework/core#207 has been merged into `master`, which is 10.0.0-dev
